### PR TITLE
application service discovery support etcd reporter

### PR DESCRIPTION
**Pull request by WilliamLeaves**
**What this PR does**:
1 application service discovery support etcd reporter
2 fix the comments of PublishAppMetadata and GetAppMetadata

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: metadata/report/etcd/report.go [hunk 1]
```diff
@@ -52,14 +53,28 @@ type etcdMetadataReport struct {
...
-	// TODO will implement
-	panic("implement me")
+	key := e.getNodeKey(metadataIdentifier)
+	_data, err := e.client.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	info := &common.MetadataInfo{}
+	err = json.Unmarshal([]byte(_data), info)
+	if err != nil {
```
Comment (hxmhlt) [inline metadata/report/etcd/report.go:63]: Why does it start with ’_‘ ？Is it golang's style?.. ：）

**Review by gaoxinge — APPROVED** [review]

File: metadata/report/etcd/report_test.go [hunk 2]
```diff
@@ -94,8 +94,14 @@ func TestEtcdMetadataReport_CRUD(t *testing.T) {
 	err = metadataReport.SaveSubscribedData(subMi, string(urls))
 	assert.Nil(t, err)
 
+	metadataInfo := common.NewMetadataInfWithApp(subMi.Application)
 	err = metadataReport.RemoveServiceMetadata(serviceMi)
 	assert.Nil(t, err)
+	err = metadataReport.PublishAppMetadata(subMi, metadataInfo)
+	assert.Nil(t, err)
+
+	_, err = metadataReport.GetAppMetadata(subMi)
```
Comment (xavier-niu) [inline metadata/report/etcd/report_test.go:103]: Why don't check the return value of `GetAppMetadata` here?

**Review by LaurenceLiZhixin — APPROVED** [review]

File: metadata/report/etcd/report.go (hunk 1 above)
Comment (WilliamLeaves) [reply metadata/report/etcd/report.go:63]: It should be "data". Starting with '_' is just a self-style when naming []byte type variables

File: metadata/report/etcd/report_test.go (hunk 2 above)
Comment (WilliamLeaves) [reply metadata/report/etcd/report_test.go:103]: Yes, It should be...

File: metadata/report/etcd/report.go [hunk 3]
```diff
@@ -52,14 +53,28 @@ type etcdMetadataReport struct {
...
 func (e *etcdMetadataReport) GetAppMetadata(metadataIdentifier *identifier.SubscriberMetadataIdentifier) (*common.MetadataInfo, error) {
-	// TODO will implement
-	panic("implement me")
+	key := e.getNodeKey(metadataIdentifier)
+	data, err := e.client.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	info := &common.MetadataInfo{}
+	err = json.Unmarshal([]byte(data), info)
```
Comment (AlexStocks) [inline metadata/report/etcd/report.go:62]: pls delete the following 4 line codes. `return info, err` is just enough.

File: metadata/report/etcd/report.go [hunk 4]
```diff
@@ -52,14 +53,28 @@ type etcdMetadataReport struct {
...
+		return nil, err
+	}
+	return info, nil
 }
 
 // PublishAppMetadata publish metadata info to etcd
 func (e *etcdMetadataReport) PublishAppMetadata(metadataIdentifier *identifier.SubscriberMetadataIdentifier, info *common.MetadataInfo) error {
-	// TODO will implement
-	panic("implement me")
+	key := e.getNodeKey(metadataIdentifier)
```
Comment (AlexStocks) [inline metadata/report/etcd/report.go:71]:
```go
	key := e.getNodeKey(metadataIdentifier)
	value, err := json.Marshal(info)
	if err == nil {
	     err = e.client.Put(key, string(value))
	}
	return err
```

**Review by AlexStocks — APPROVED** [review]

**Pull request merged**